File: binding-csharp/runtime/src/test/csharp/Etch/Util/TestStrIntHashMapSerializer.cs
// $Id$
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.
// 
using System;
using Org.Apache.Etch.Bindings.Csharp.Msg;
using Org.Apache.Etch.Bindings.Csharp.Support;
using NUnit.Framework;

namespace Org.Apache.Etch.Bindings.Csharp.Util
{
    [TestFixture]
    public class TestStrIntHashMapSerializer
    {
        private readonly ValueFactory vf = new DummyValueFactory();

        [TestFixtureSetUp]
        public void First()
        {
            Console.WriteLine();
            Console.Write( "TestStrIntHashMapSerializer" );
        }


        [Test]
        public void TestImport0() 
	    {
		    StrIntHashMap map = new StrIntHashMap();
		    TestMap( map );
	    }

        [Test]
	    public void TestImport1() 
	    {
		    StrIntHashMap map = new StrIntHashMap();
		    map.Add( "a", 1 );
		    TestMap( map );
	    }

        [Test]
	    public void TestImport2() 
	    {
		    StrIntHashMap map = new StrIntHashMap();
		    map.Add( "a", 1 );
		    map.Add( "b", 2 );
		    TestMap( map );
	    }

        [Test]
	    public void TestImport3() 
	    {
		    StrIntHashMap map = new StrIntHashMap();
		    map.Add( "a", 1 );
		    map.Add( "b", 2 );
		    map.Add( "c", 3 );
		    TestMap( map );
	    }
    	
        [Test]
	    public void TestImportNull1() 
	    {
		    StrIntHashMap map = new StrIntHashMap();
            map.Add( "a", 1 );
		    map.Add( "b", null );
            map.Add( "c", 2 );
		    TestMap( map );
	    }
    	
	    // no tests for null keys as string? isn't allowed

        public void TestMap( StrIntHashMap map )
        {
            XType type = new XType( "map" );
		  
            Class2TypeMap class2type = new Class2TypeMap();
		    StrIntHashMapSerializer.Init( type, class2type );
		    ImportExportHelper helper = type.GetImportExportHelper();
    		
		    StructValue sv = helper.ExportValue( vf, map );
    		
		    Assert.AreEqual( sv.GetXType, type );
    		
		    StrIntHashMap map2 = (StrIntHashMap) helper.ImportValue( sv );
    		
		    Assert.AreEqual( map, map2 );
        }

    }
}

File: binding-java/runtime/src/test/java/org/apache/etch/bindings/java/util/TestStrIntHashMapSerializer.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.bindings.java.util;

import static org.junit.Assert.assertEquals;

import org.apache.etch.bindings.java.msg.ImportExportHelper;
import org.apache.etch.bindings.java.msg.StructValue;
import org.apache.etch.bindings.java.msg.Type;
import org.apache.etch.bindings.java.msg.ValueFactory;
import org.apache.etch.bindings.java.support.Class2TypeMap;
import org.apache.etch.bindings.java.support.DummyValueFactory;
import org.junit.Test;


/** Tests StrIntHashMapSerializer */
public class TestStrIntHashMapSerializer
{
	private final ValueFactory vf = new DummyValueFactory();
	
	/** @throws Exception */
	@Test
	public void test0() throws Exception
	{
		StrIntHashMap map = new StrIntHashMap();
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void test1() throws Exception
	{
		StrIntHashMap map = new StrIntHashMap();
		map.put( "a", 1 );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void test2() throws Exception
	{
		StrIntHashMap map = new StrIntHashMap();
		map.put( "a", 1 );
		map.put( "b", 2 );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void test3() throws Exception
	{
		StrIntHashMap map = new StrIntHashMap();
		map.put( "a", 1 );
		map.put( "b", 2 );
		map.put( "c", 3 );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void testNullValue() throws Exception
	{
		StrIntHashMap map = new StrIntHashMap();
		map.put( "a", 1 );
		map.put( "b", null );
		map.put( "c", 3 );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void testNullKey() throws Exception
	{
		StrIntHashMap map = new StrIntHashMap();
		map.put( "a", 1 );
		map.put( null, 2 );
		map.put( "c", 3 );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void testNullKeyValue() throws Exception
	{
		StrIntHashMap map = new StrIntHashMap();
		map.put( "a", 1 );
		map.put( null, null );
		map.put( "c", 3 );
		testMap( map );
	}
	
	private void testMap( StrIntHashMap map )
	{
//		System.out.println( "---------------------------------------------" );
//		System.out.println( "map = "+map );
		
		Type type = new Type( "map" );
		Class2TypeMap class2type = new Class2TypeMap();
		StrIntHashMapSerializer.init( type, class2type );
		ImportExportHelper helper = type.getImportExportHelper();
		
		StructValue sv = helper.exportValue( vf, map );
//		System.out.println( "sv = "+sv );
		
		assertEquals( sv.type(), type );
		
		StrIntHashMap map2 = (StrIntHashMap) helper.importValue( sv );
//		System.out.println( "map2 = "+map2 );
		
		assertEquals( map, map2 );
	}
}
